lib/seattleflu/id3c/cli/command/backfill_sample_ids.py

```diff
@@ -0,0 +1,51 @@
+"""
+Backfill sample ids
+
+Mint new sample identifiers and assign them to warehouse.sample records that are missing sample identifiers. This is useful for tiny swabs samples, which do not have room on the tube for a sample identifier, only a collection identifier. This process fills in the sample record in our warehouse table but notably does not notify the lab of any identifier assignments. We can use this backfilled sample ids as an identifier when sharing data with third parties.
+"""
+
+import logging
+
+from id3c.cli import cli
+from id3c.cli.command import DatabaseSessionAction, with_database_session
+from id3c.db import mint_identifiers, upsert_sample
+from id3c.db.session import DatabaseSession
+
+LOG = logging.getLogger(__name__)
+
+
+@cli.command("backfill-sample-ids", help = __doc__)
+@with_database_session(pass_action = True)
+def backfill_sample_ids(*, db: DatabaseSession, action: DatabaseSessionAction):
+    LOG.debug(f"Backfilling sample ids")
+
+    collection_identifiers = db.fetch_all("""
+            select sample.collection_identifier
+            from warehouse.sample
+            join warehouse.identifier on sample.collection_identifier = identifier.uuid::text
+            join warehouse.identifier_set using (identifier_set_id)
+            where identifier is null and
+            identifier_set.name in ('collections-uw-tiny-swabs-home', 'collections-uw-tiny-swabs-observed')
+            """)
+
+    if not collection_identifiers:
+        LOG.info(f"No sample records that need sample id backfill")
+        return
+
+    LOG.info(f"Fetched {len(collection_identifiers):,} sample records that need sample id backfill")
+
+    minted_sample_identifiers = mint_identifiers(db, 'samples', len(collection_identifiers))
+    assert len(minted_sample_identifiers) == len(collection_identifiers), "Didn't generate appropriate number of identifiers"
+
+    for sample_identifier, collection_identifier in zip(minted_sample_identifiers, collection_identifiers):
+        LOG.info(f"Assigning sample id {sample_identifier.uuid} to collection id {collection_identifier.collection_identifier} in warehouse.samples")
+
+        upsert_sample(db,
+            update_identifiers= True,
+            identifier = sample_identifier.uuid,
+            collection_identifier = collection_identifier.collection_identifier,
+            collection_date = None,
+            encounter_id = None,
+            additional_details = {}
+        )
```

Comment:
Works great! My only thought is that `upsert_sample` may be overkill here since it's going to run additional queries to check whether the record exists before it runs the update query. In this case you already have confirmed the records exist, so could get away with a simple `cursor.execute("update ...")`